### PR TITLE
Close dropdowns when user clicks outside of them (fixes #5490)

### LINKED ISSUE #5490: Clicking away from menu should close it
Several times recently, I've opened up one of the pulldown menus (Social, the settings gear, etc.), and tried reflexively to dismiss it by clicking in empty space somewhere. That closes modals in Habit and similar menus in many other apps, but here it has no effect. It takes a second of puzzlement before I remember I have to click the menu itself again to close it.

Getting a click-away dismiss would be nice and spare me a few senior moments.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/23049940-clicking-away-from-menu-should-close-it?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by keysmashes**
Fixes #5490

### Changes

Makes dropdowns (user, social, help, notifications, audio theme etc.) close when the user clicks outside of them. This is done by changing the dropdowns to [ui-bootstrap][uib] dropdowns.

Since the full-screen mobile dropdowns are a hack rather than proper modals, there is an unintended behaviour change here – previously, the close button in the bottom-right was the only way to close the dropdown without clicking on a link, whereas now any click anywhere on the screen will dismiss the dropdown. I don't _think_ there's an easy way to fix this.

Note that there are many things that could be done to clean this up more; I've done essentially the minimum to make the change happen. The directives(?) for the original dropdown toggle method still exist, as well as many unnecessary `data-close-menu`s. If I can avoid removing them, I will (since I think the new client will make this code obsolete, and I'm lazy), but I'm happy to remove all the pieces of the original dropdowns if required.

[uib]: https://angular-ui.github.io/bootstrap/versioned-docs/0.13.0/#/dropdown

#### Before

image

image

image

#### After

image

image

image

----
UUID: 0e1d15f7-f36b-4f15-a9e1-0c31b732dac9

### CODE REVIEW

**Comment by Alys** [comment]
I've put this on my test server and tried it on desktop and mobile. It seems good. The behaviour on mobile actually seems the same to me as on the production site - tapping anywhere on the production site will close a menu too; I don't think that needs to be changed.

Based on the behaviour I've seen in my tests, I'd be happy for this to go live. However I haven't reviewed the code.

If anyone else wants to try it on my test server, it's here: http://192.241.249.166:3010/
I don't guarantee that site will be up all the time though. :)

File: website/views/shared/header/menu.jade [hunk 1]
```diff
@@ -99,14 +99,14 @@ nav.toolbar(ng-controller='MenuCtrl')
...
+      li.toolbar-button-dropdown(dropdown)
         a(ui-sref='options.social.inbox', ng-if='user.inbox.newMessages')
           span.badge.badge-danger {{user.inbox.newMessages}}
-        a(ui-sref='options.social.tavern', data-close-menu)
+        a(ui-sref='options.social.tavern')
           span=env.t('social')
-        a(ng-class='{active: _expandedMenu.menu === "social"}', data-expand-menu, menu='social')
+        a(dropdown-toggle)
           span &#9776;
-        div(ng-show='_expandedMenu.menu === "social"', data-close-menu)
```
Comment (paglias) [inline website/views/shared/header/menu.jade:109]: All this references to `_expandedMenu ` only exist in this file or can be removed from somewhere else too?

Comment (keysmashes) [reply website/views/shared/header/menu.jade:109]: It looks like there are references in the close-menu and expand-menu directives (unsurprisingly) and the tests for those directives, and I missed a couple of references in this file on lines [292](https://github.com/HabitRPG/habitica/pull/8657/files/19b9be27a55a6f69c88d538f4d560be298161e4a#diff-0d726286f04628feed0f89ff02f1078bR292) and 296. I'll remove the directives and tests, I guess?

What happens if the `ul.toolbar-bailey` on lines 292/296 have the `inactive` class? With this PR, they never will, but I don't know how to test that or how to see what they do.

Comment (paglias) [reply website/views/shared/header/menu.jade:109]: Not sure, it seems they always have the inactive class right now but that doesn't change anything in the CSS

**Comment by keysmashes** [comment]
I've now removed most of the traces of the original dropdowns. The only parts that are left (as far as I can tell) are on lines [292][] and [296][], since I don't know what the `inactive` class on those elements is meant to do -- I'm happy to remove or leave the class, although it would probably be best to remove the references to `_expandedMenu` to avoid confusing people in the future.

[292]: https://github.com/HabitRPG/habitica/pull/8657/files#diff-0d726286f04628feed0f89ff02f1078bL292
[296]: https://github.com/HabitRPG/habitica/pull/8657/files#diff-0d726286f04628feed0f89ff02f1078bL296

**Comment by paglias** [comment]
As we're part through a rewrite of the front end I think this is okay for now

**Comment by khipkin** [comment]
Wasn't it mentioned by someone (maybe @Alys?) that this might be an accessibility issue?

**Comment by Alys** [comment]
We did talk about accessibility issues related to the menus in the past but the biggest concern was with menus closing automatically once you moved off them. This PR only closes them when you click which I think is a good compromise.

**Comment by TheHollidayInn** [comment]
Noted towards your Tier! Thanks :D

**Pull request merged**